Expand the documentation for vi

Yilin-Yang opened the pull request:
Add explanations of:
-Why vim is so popular among programmers.
-Some of the culture surrounding vim (popular vim plugin developers,
some common jokes among vim users, etc.)
-The purpose of the .vimrc, how to version control it, and how
to better organize it by splitting it into multiple smaller files.
-What vimscript is (though not necessarily how to use it).
-Plugin managers (pathogen.vim, Vundle).
-vim and the neovim fork.
-How to configure (neo)vim to work like an IDE.

Add lists of:
-Some of vim's coolest, built-in features. Link to vimcasts.org, the vim
wikia page, and some other webpages explaining how macros,
search-and-replace, vim regex, etc. work.
-Plugin recommendations (i.e. git clone https://github.com/tpope/literally-everything)

Also add a few .GIF files screencasting macro usage, vim's toggle-case,
and visual-block copy-paste.

# Advanced HW 5 submission

Closes #350 

Name: Yilin Yang
Uniqname: yiliny

cameron-gagnon commented:
Wow! This looks awesome! 🎊  🎉 

I've gone through it and it looks pretty rich with info. However, and I almost feel bad for asking, but I noticed there wasn't any info on the different [VIM modes](https://en.wikibooks.org/wiki/Learning_the_vi_Editor/Vim/Modes)? If you feel compelled to add a few snippets on that, that'd really bring this to a 120/100. Otherwise, I'd still be fine giving credit and closing this out.

Let me know what you think! :) Thanks for adding this gem! I even learned a few things and found some good resources you linked to.

Yilin-Yang commented:
Absolutely! It'll probably take me a little bit, but modal editing is definitely a big part of using vim, so I'd be happy to add it in! Professor Darden also showed me how to use asciinema, so I should be able to update the screencasts as well!

I'll let you guys know when the changes are ready!

Yilin-Yang commented:
Finished the sections on modal editing! They cover insert mode, normal mode, and visual mode (ordinary/visual line/visual block). I also replaced the old GIF screencasts with GIFs generated using asciinema and [asciinema/asciicast2gif](https://github.com/asciinema/asciicast2gif); added GIFs demonstrating insert, normal, and visual mode; and also made the GIFs adaptively resizeable on differently-sized browser windows!

It seems like that last change has the side effect of making the CreativeCommons logo at the bottom of the page gigantic, however. Not entirely sure how to fix that. :( There's also some weirdness going on with my course schedule right now, so I won't be asking for homework credit for this.

Let me know if there's anything else! :)
